### PR TITLE
Batch hmac - (#5850)

### PR DESCRIPTION
**Pull request by martinwaite**
Guidance is needed on some points:

1. batch_input added to pathHMACwrite.  Should I extend this to pathHMACverify ? would that extend to path_sign_verify ?
2. There are several calls to p.Unlock() before returns:  is there a reason not to use defer p.Unlock() ?
3. Test is shallow - returned values should be checked - perhaps using verify ?
4. batch_input only supports "input" parameters. Should I include other parameters ?

Most of the changes are copied from batch_input in path_encrypt.  I was surprised that the framework.Path structure did not mention batch_input.  Why is it not there ?

### CODE REVIEW

**Comment by martinwaite** [comment]
For the time being, I have reverted to the existing behaviour of accepting requests without an input field and returning the HMAC of a blank field.  

I guess there would be no harm in fixing up the path_config_test case by copying the `plaintext` field into the `input` field for the HMAC tests.

In the batch_response, hmac fields are uppercase HMAC rather than lowercase for non-batch.  I will fix that later.

**Comment by martinwaite** [comment]
I have fixed up the transit/path_config_test so that `input` is supplied as well as `plaintext`.  This allows the test to pass with stricter parameter checking in-place.

I have reverted transit/path_hmac pathHMACWrite to return an error if no input is supplied.  When batch_input is supplied, any item in the data array without an `input` field will receive an error message in the response.  For example:

```
{
  "batch_input": [
    {
      "input": "adba32=="
    },
    {
      "input": "adba32=="
    },
    {},
    {
      "input": ""
    }
  ]
}
```

Now returns:
```
{
  "request_id": "2d15cdc3-c26c-8259-b0c4-7251371dbaf8",
  "lease_id": "",
  "renewable": false,
  "lease_duration": 0,
  "data": {
    "batch_results": [
      {
        "hmac": "vault:v1:1jFhRYWHiddSKgEFyVRpX8ieX7UU+748NBwHKecXE3hnGBoAxrfgoD5U0yAvji7b5X6V1fP
      },
      {
        "hmac": "vault:v1:1jFhRYWHiddSKgEFyVRpX8ieX7UU+748NBwHKecXE3hnGBoAxrfgoD5U0yAvji7b5X6V1fP
      },
      {
        "error": "missing input for HMAC"
      },
      {
        "hmac": "vault:v1:/wsSP6iQ9ECO9RRkefKLXey9sDntzSjoiW0vBrWfUsYB0ISroyC6plUt/jN7gcOv9O+Ecow
      }
    ]
  },
  "wrap_info": null,
  "warnings": null,
  "auth": null
}
```

**Comment by briankassouf** [comment]
Hi @martinwaite, thanks for contributing this change! We'd like to get this merged in soon, so we were wondering if you could add API docs for the new API?

**Comment by martinwaite** [comment]
Hi @briankassouf - I have added some api docs.

File: builtin/logical/transit/path_hmac.go [hunk 1]
```diff
@@ -116,31 +128,88 @@ func (b *backend) pathHMACWrite(ctx context.Context, req *logical.Request, d *fr
...
+
+		rawInput, ok := item["input"]
+		if !ok {
+			batchResponseItems[i].Error = "missing input for HMAC"
+			continue
+		}
+
+		input, err := base64.StdEncoding.DecodeString(rawInput)
+		if err != nil {
+			batchResponseItems[i].Error = fmt.Sprintf("unable to decode input as base64: %s", err)
```
Comment (briankassouf) [inline builtin/logical/transit/path_hmac.go]: I think there is a missing `continue` here

Comment (martinwaite) [reply builtin/logical/transit/path_hmac.go]: You are correct.  

I have added another input to the test with invalid base64 data to exercise this. 

In practice, it turns out that missing the `continue` had no impact on the output and an error is generated either with or without the `continue`.  However, adding the `continue` avoids needlessly generating an HMAC of the empty input string - so let's do that.

**Comment by briankassouf** [comment]
@martinwaite After thinking about it some more it would be really nice to also offer a batch verify and a batch sign API at the same time as this change. Is that something you'd be willing to add to this PR? I imagine the changeset for verify/sign would look very similar to what you did for hmac.

**Comment by martinwaite** [comment]
@briankassouf - I will extend the PR to cover sign and verify.  I should get it done over the weekend.

**Comment by martinwaite** [comment]
Refined the error-handling so that when batch mode simulates simple 'input' case, errors are reported in same way as before batch was added.  This is working for the Sign path, but needs to be extended to the verify and hmac cases.

**Comment by martinwaite** [comment]
@briankassouf - This is ready for review now.

**Comment by martinwaite** [comment]
@briankassouf - I need help with the failing tests.  5 of the tests are something to do with the documentation.  The logs show a ruby stack trace - I have no idea how to deal with that.  One of the CI builds tripped over a JWT issue - I can't tell if that is something I did.

**Comment by martinwaite** [comment]
@briankassouf - all tests are passing now.  This is ready for review.

File: builtin/logical/transit/path_hmac.go [hunk 2]
```diff
@@ -116,70 +136,105 @@ func (b *backend) pathHMACWrite(ctx context.Context, req *logical.Request, d *fr
 		return nil, fmt.Errorf("HMAC key value could not be computed")
 	}
 
-	var hf hash.Hash
+	var hash_alg func() hash.Hash
```
Comment (kalafut) [inline builtin/logical/transit/path_hmac.go]: Prefer `hashAlg`

File: builtin/logical/transit/path_hmac.go [hunk 3]
```diff
@@ -195,49 +250,150 @@ func (b *backend) pathHMACVerify(ctx context.Context, req *logical.Request, d *f
...
-		p.Unlock()
-		return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
-	}
-	if key == nil {
-		p.Unlock()
-		return nil, fmt.Errorf("HMAC key value could not be computed")
-	}
-
-	var hf hash.Hash
+	var hash_alg func() hash.Hash
```
Comment (kalafut) [inline builtin/logical/transit/path_hmac.go]: Prefer `hashAlg`

File: builtin/logical/transit/path_sign_verify.go [hunk 4]
```diff
@@ -5,12 +5,57 @@ import (
...
+// A map type allows us to distinguish between empty and missing values.
+type batchRequestSignItem map[string]string
+
+// BatchResponseSignItem represents a response item for batch processing
+type batchResponseSignItem struct {
+	// signature for the input present in the corresponding batch
+	// request item
+	Signature string `json:"signature,omitempty" structs:"signature" mapstructure:"signature"`
+
+	PublicKey []byte `json:"publickey,omitempty" structs:"publickey" mapstructure:"publckey"`
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify.go]:
```suggestion
	PublicKey []byte `json:"publickey,omitempty" structs:"publickey" mapstructure:"publickey"`
```

File: builtin/logical/transit/path_sign_verify.go [hunk 5]
```diff
@@ -222,63 +261,175 @@ func (b *backend) pathSignWrite(ctx context.Context, req *logical.Request, d *fr
...
+			continue
+		}
+
+		if p.Type.HashSignatureInput() && !prehashed {
+			var hf = keysutil.HashFuncMap[hashAlgorithm]()
+			hf.Write(input)
+			input = hf.Sum(nil)
+		}
+
+		contextRaw, _ := item["context"]
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify.go]:
```suggestion
		contextRaw := item["context"]
```

File: builtin/logical/transit/path_sign_verify.go [hunk 6]
```diff
@@ -222,63 +261,175 @@ func (b *backend) pathSignWrite(ctx context.Context, req *logical.Request, d *fr
...
-		return b.pathHMACVerify(ctx, req, d, hmac)
+	case sigFound && hmacFound:
+		return logical.ErrorResponse("elements of batch_input must all provide 'signature' or all provide 'hmac'"), logical.ErrInvalidRequest
+
+	case missing && sigFound:
+		return logical.ErrorResponse("some elements of batch_input are missing 'signature'"), logical.ErrInvalidRequest
+
+	case missing && hmacFound:
+		return logical.ErrorResponse("some elements of batch_input are missing 'hmac'"), logical.ErrInvalidRequest
+
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify.go]: Do you also need `case missing:`?  i.e. no elements provide either 'signature' or 'hmac'

File: builtin/logical/transit/path_sign_verify.go [hunk 7]
```diff
@@ -326,42 +472,82 @@ func (b *backend) pathVerifyWrite(ctx context.Context, req *logical.Request, d *
...
-		default:
-			p.Unlock()
-			return nil, err
+		if p.Type.HashSignatureInput() && !prehashed {
+			hf := keysutil.HashFuncMap[hashAlgorithm]()
+			hf.Write(input)
+			input = hf.Sum(nil)
+		}
+
+		contextRaw, _ := item["context"]
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify.go]:
```suggestion
		contextRaw := item["context"]
```

File: builtin/logical/transit/path_sign_verify.go [hunk 8]
```diff
@@ -222,63 +261,175 @@ func (b *backend) pathSignWrite(ctx context.Context, req *logical.Request, d *fr
...
 
 	p.Unlock()
 	return resp, nil
 }
 
 func (b *backend) pathVerifyWrite(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	sig := d.Get("signature").(string)
-	hmac := d.Get("hmac").(string)
+	batchInputRaw := d.Raw["batch_input"]
+	var batchInputItems []batchRequestSignItem
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify.go]:
```suggestion
	var batchInputItems []batchRequestVerifyItem
```

File: builtin/logical/transit/path_sign_verify_test.go [hunk 9]
```diff
@@ -274,47 +329,85 @@ func TestTransit_SignVerify_ED25519(t *testing.T) {
...
 			t.Fatal("expected non-nil response")
 		}
 		if errExpected {
-			if !resp.IsError() {
-				t.Fatalf("bad: got error response: %#v", *resp)
+			if resp.IsError() {
+				return nil
 			}
-			return ""
+			t.Fatalf("bad: got error response: %#v", *resp)
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify_test.go]:
```suggestion
			t.Fatalf("bad: expected error response, got: %#v", *resp)
```

File: builtin/logical/transit/path_sign_verify_test.go [hunk 10]
```diff
@@ -274,47 +329,85 @@ func TestTransit_SignVerify_ED25519(t *testing.T) {
...
 		// memoize any pubic key
 		if key, ok := resp.Data["public_key"]; ok {
 			req.Data["public_key"] = key
 		}
-		return value.(string)
+		// batch_input supplied
+		if _, ok := req.Data["batch_input"]; ok {
+			batchRequestItems := req.Data["batch_input"].([]batchRequestSignItem)
+
+			batch_results, ok := resp.Data["batch_results"]
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify_test.go]: Prefer `batchResults`

File: builtin/logical/transit/path_sign_verify_test.go [hunk 11]
```diff
@@ -274,47 +329,85 @@ func TestTransit_SignVerify_ED25519(t *testing.T) {
...
+			req.Data["signature"] = sig[0]
+		}
 		resp, err := b.HandleRequest(context.Background(), req)
 		if err != nil && !errExpected {
 			t.Fatalf("got error: %v, sig was %v", err, sig)
 		}
 		if errExpected {
 			if resp != nil && !resp.IsError() {
-				t.Fatalf("bad: got error response: %#v", *resp)
+				t.Fatalf("bad: got error response: %#v\n%#v", *resp, req)
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify_test.go]:
```suggestion
				t.Fatalf("bad: expect error response, got: %#v\n%#v", *resp, req)
```

File: builtin/logical/transit/path_sign_verify_test.go [hunk 12]
```diff
@@ -324,52 +417,57 @@ func TestTransit_SignVerify_ED25519(t *testing.T) {
 		if resp.IsError() {
 			t.Fatalf("bad: got error response: %#v", *resp)
 		}
+
+		// batch_input field supplied
+		if _, ok := req.Data["batch_input"]; ok {
+			batchRequestItems := req.Data["batch_input"].([]batchRequestSignItem)
+
+			batch_results, ok := resp.Data["batch_results"]
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify_test.go]: Prefer `batchResults`

File: website/source/api/secret/transit/index.html.md [hunk 13]
```diff
@@ -747,13 +747,26 @@ be used.
 
 - `input` `(string: <required>)` – Specifies the **base64 encoded** input data.
```
Comment (kalafut) [inline website/source/api/secret/transit/index.html.md]: This parameter is no longer <required>.

File: website/source/api/secret/transit/index.html.md [hunk 14]
```diff
@@ -806,6 +870,28 @@ supports signing.
...
```
Comment (kalafut) [inline website/source/api/secret/transit/index.html.md]: This is no longer a required parameter.

**Comment by kalafut** [comment]
@martinwaite This looks really good! I left a few minor comments but overall I think it's close.

@briankassouf Any thoughts on having a limit (possible pretty high) on the max batch size? I'm wondering if having no limit might lure users into dumping all items into a batch, and then as their batch size grows they might start hitting request timeout errors, or Vault has issues, etc. Enforcing some level of batching on the caller's side now might avoid those issues.

**Comment by jefferai** [comment]
@kalafut we already have a per-listener maximum request size (with a default of about 32 MB). I guess the question would be, in a normal request timeout period, what does this translate to in terms of max batched operations. But that request timeout is also driven by the client, so if the client is hitting errors it can change it...

File: builtin/logical/transit/path_hmac.go (hunk 2 above)
Comment (martinwaite) [reply builtin/logical/transit/path_hmac.go]: ok

File: builtin/logical/transit/path_hmac.go (hunk 3 above)
Comment (martinwaite) [reply builtin/logical/transit/path_hmac.go]: ok

File: builtin/logical/transit/path_sign_verify.go (hunk 4 above)
Comment (martinwaite) [reply builtin/logical/transit/path_sign_verify.go]: oops - ok

File: builtin/logical/transit/path_sign_verify.go (hunk 6 above)
Comment (martinwaite) [reply builtin/logical/transit/path_sign_verify.go]: added case for missing on its own

File: builtin/logical/transit/path_sign_verify.go (hunk 8 above)
Comment (martinwaite) [reply builtin/logical/transit/path_sign_verify.go]: oops - ok

File: website/source/api/secret/transit/index.html.md (hunk 13 above)
Comment (martinwaite) [reply website/source/api/secret/transit/index.html.md]: I have changed "<required>" to nil - and added a sentence about either input or batch_input being needed.   I'm not sure about "nil" being correct .

File: website/source/api/secret/transit/index.html.md (hunk 14 above)
Comment (martinwaite) [reply website/source/api/secret/transit/index.html.md]: I have changed "<required>" to nil - and added a sentence about either input or batch_input being needed. I'm not sure about "nil" being correct .

**Comment by martinwaite** [comment]
@kalafut - I have implemented all of your code suggestions.  I'm not sure about the changes I made to the website doc - please check.

File: builtin/logical/transit/path_sign_verify_test.go [hunk 15]
```diff
@@ -274,47 +328,85 @@ func TestTransit_SignVerify_ED25519(t *testing.T) {
...
 			t.Fatal("expected non-nil response")
 		}
 		if errExpected {
-			if !resp.IsError() {
-				t.Fatalf("bad: got error response: %#v", *resp)
+			if resp.IsError() {
+				return nil
 			}
-			return ""
+			t.Fatalf("bad: expected response, got: %#v", *resp)
```
Comment (kalafut) [inline builtin/logical/transit/path_sign_verify_test.go]:
```suggestion
			t.Fatalf("bad: expected error response, got: %#v", *resp)
```

File: website/source/api/secret/transit/index.html.md [hunk 16]
```diff
@@ -745,15 +745,29 @@ be used.
     - `sha2-384`
     - `sha2-512`
 
-- `input` `(string: <required>)` – Specifies the **base64 encoded** input data.
+- `input` `(string: nil)` – Specifies the **base64 encoded** input data. One of
```
Comment (kalafut) [inline website/source/api/secret/transit/index.html.md]:
```suggestion
- `input` `(string: "")` – Specifies the **base64 encoded** input data. One of 
```

File: website/source/api/secret/transit/index.html.md [hunk 17]
```diff
@@ -805,7 +870,30 @@ supports signing.
...
```
Comment (kalafut) [inline website/source/api/secret/transit/index.html.md]:
```suggestion
- `input` `(string: "")` – Specifies the **base64 encoded** input data. One of 
```

File: website/source/api/secret/transit/index.html.md [hunk 18]
```diff
@@ -894,6 +1023,31 @@ data.
   `/transit/hmac` function. Either this must be supplied or `signature` must be
   supplied.
 
+- `batch_input` `(array<object>: nil)` – Specifies a list of items for processing.
```
Comment (kalafut) [inline website/source/api/secret/transit/index.html.md]: (GH won't let me add this comment outside the diff range).  Please update the `input` docs here as you did the others.

File: builtin/logical/transit/path_hmac.go [hunk 19]
```diff
@@ -116,70 +136,105 @@ func (b *backend) pathHMACWrite(ctx context.Context, req *logical.Request, d *fr
 		return nil, fmt.Errorf("HMAC key value could not be computed")
 	}
 
-	var hf hash.Hash
+	var hashAlg func() hash.Hash
```
Comment (jefferai) [inline builtin/logical/transit/path_hmac.go]: This should probably use the keysutil.HashTypeMap stuff we are using elsewhere in `transit`

File: builtin/logical/transit/path_hmac.go [hunk 20]
```diff
@@ -195,49 +250,150 @@ func (b *backend) pathHMACVerify(ctx context.Context, req *logical.Request, d *f
...
-		p.Unlock()
-		return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
-	}
-	if key == nil {
-		p.Unlock()
-		return nil, fmt.Errorf("HMAC key value could not be computed")
-	}
-
-	var hf hash.Hash
+	var hashAlg func() hash.Hash
```
Comment (jefferai) [inline builtin/logical/transit/path_hmac.go]: Same comment

File: builtin/logical/transit/path_sign_verify.go [hunk 21]
```diff
@@ -5,12 +5,57 @@ import (
...
 
+// BatchRequestSignItem represents a request item for batch processing.
+// A map type allows us to distinguish between empty and missing values.
+type batchRequestSignItem map[string]string
+
+// BatchResponseSignItem represents a response item for batch processing
+type batchResponseSignItem struct {
+	// signature for the input present in the corresponding batch
+	// request item
+	Signature string `json:"signature,omitempty" structs:"signature" mapstructure:"signature"`
```
Comment (jefferai) [inline builtin/logical/transit/path_sign_verify.go]: Here and elsewhere, if you are not using the `structs` package in your tests for this functionality, please don't decorate it. We are generally trying to get rid of usage of that package.

**Comment by kalafut** [comment]
@martinwaite Thanks for addressing the comments and the contribution as a whole. There a few additional comments that have been left, but at this point I'm going to merge your PR and make those changes.

**Pull request merged**